File: app/src/main/java/com/itis/android/lessontwo/ui/comics/ComicsActivity.java
package com.itis.android.lessontwo.ui.comics;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.v7.widget.Toolbar;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.PresenterType;
import com.itis.android.lessontwo.R;
import com.itis.android.lessontwo.model.comics.Comics;
import com.itis.android.lessontwo.model.comics.ComicsTextObject;
import com.itis.android.lessontwo.ui.base.BaseActivity;
import com.itis.android.lessontwo.utils.ImageLoadHelper;

import static com.itis.android.lessontwo.utils.Constants.ID_KEY;
import static com.itis.android.lessontwo.utils.Constants.NAME_KEY;

/**
 * Created by Nail Shaykhraziev on 25.02.2018.
 */
public class ComicsActivity extends BaseActivity implements ComicsView {

    private CollapsingToolbarLayout collapsingToolbar;
    private Toolbar toolbar;
    private ImageView ivCover;
    private TextView tvDescription;
    private TextView tvPrice;
    private TextView tvPages;

    @InjectPresenter
    ComicsPresenter presenter;

    private long id;

    public static void start(@NonNull Activity activity, @NonNull Comics comics) {
        Intent intent = new Intent(activity, ComicsActivity.class);
        intent.putExtra(NAME_KEY, comics.getName());
        intent.putExtra(ID_KEY, comics.getId());
        activity.startActivity(intent);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        FrameLayout contentFrameLayout = findViewById(R.id.container);
        getLayoutInflater().inflate(R.layout.activity_comics, contentFrameLayout);
        initViews();

        id = getIntent().getLongExtra(ID_KEY, 0);
    }

    @Override
    public void getComicsId() {
        presenter.init(id);
    }

    @Override
    public void setPageCount(final Comics comics) {
        tvPages.setText(String.valueOf(comics.getPageCount()));
    }

    @Override
    public void setPrice(final Comics comics) {
        if (comics.getPrices() != null && !comics.getPrices().isEmpty()) {
            tvPrice.setText(getString(R.string.price_format, String.valueOf(comics.getPrices().get(0).getPrice())));
        }
    }

    @Override
    public void setDescription(final Comics comics) {
        if (comics.getTextObjects() != null) {
            StringBuilder description = new StringBuilder();
            for (ComicsTextObject comicsTextObject : comics.getTextObjects()) {
                description.append(comicsTextObject.getText()).append("\n");
            }
            tvDescription.setText(description.length() > 0 ?
                    description.toString().trim() : getString(R.string.text_desc_not_found));
        }
    }

    @Override
    public void setImage(final Comics comics) {
        if (comics.getImage() != null) {
            ImageLoadHelper.loadPicture(ivCover, String.format("%s.%s", comics.getImage().getPath(),
                    comics.getImage().getExtension()));
        } else {
            ImageLoadHelper.loadPictureByDrawable(ivCover, R.drawable.image_error_marvel_logo);
        }
    }

    @Override
    public void handleError(Throwable error) {
        Toast.makeText(this, error.getMessage(), Toast.LENGTH_SHORT).show();
    }

    private void initViews() {
        findViews();
        supportActionBar(toolbar);
        setBackArrow(toolbar);
        collapsingToolbar.setTitle(getIntent().getStringExtra(NAME_KEY));
    }

    private void findViews() {
        collapsingToolbar = findViewById(R.id.ct_comics);
        toolbar = findViewById(R.id.tb_comics);
        ivCover = findViewById(R.id.iv_comics);
        tvDescription = findViewById(R.id.tv_description);
        tvPrice = findViewById(R.id.tv_price);
        tvPages = findViewById(R.id.tv_pages);
    }
}

File: app/src/main/java/com/itis/android/lessontwo/ui/comics/ComicsActivity.kt
package com.itis.android.lessontwo.ui.comics

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.CollapsingToolbarLayout
import android.support.v7.widget.Toolbar
import android.view.View
import android.widget.*
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.PresenterType
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.itis.android.lessontwo.R
import com.itis.android.lessontwo.model.comics.Comics
import com.itis.android.lessontwo.repository.RepositoryProvider
import com.itis.android.lessontwo.ui.base.BaseActivity
import com.itis.android.lessontwo.utils.Constants.ID_KEY
import com.itis.android.lessontwo.utils.Constants.NAME_KEY
import com.itis.android.lessontwo.utils.ImageLoadHelper

class ComicsActivity : BaseActivity(), ComicsView {

    private lateinit var collapsingToolbar: CollapsingToolbarLayout
    private lateinit var toolbar: Toolbar
    private lateinit var ivCover: ImageView
    private lateinit var tvDescription: TextView
    private lateinit var tvPrice: TextView
    private lateinit var tvPages: TextView
    private lateinit var progressBar: ProgressBar

    private var id: Long = 0

    @InjectPresenter(type = PresenterType.LOCAL)
    lateinit var presenter: ComicsPresenter

    @ProvidePresenter
    fun initPresenter(): ComicsPresenter =
            ComicsPresenter(RepositoryProvider.provideComicsRepository())

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val contentFrameLayout = findViewById<FrameLayout>(R.id.container)
        layoutInflater.inflate(R.layout.activity_comics, contentFrameLayout)
        initViews()
        id = intent.getLongExtra(ID_KEY, 0)
    }

    override fun getComicsId() = presenter.init(id)

    override fun handleError(error: Throwable) =
            Toast.makeText(this, error.message, Toast.LENGTH_SHORT).show()

    override fun setPageCount(comics: Comics) {
        tvPages.text = comics.pageCount.toString()
    }

    override fun setPrice(comics: Comics) {
        if (comics.prices.isNotEmpty()) {
            tvPrice.text = getString(R.string.price_format, comics.prices[0]?.price.toString())
        }
    }

    override fun setDescription(comics: Comics) {
        if (comics.textObjects != null) {
            val description = StringBuilder()
            for (comicsTextObject in comics.textObjects) {
                description.append(comicsTextObject.text).append("\n")
            }
            tvDescription.text = if (description.isNotEmpty())
                description.toString().trim { it <= ' ' }
            else
                getString(R.string.text_desc_not_found)
        }
    }

    override fun setImage(comics: Comics) {
        if (comics.image != null) {
            ImageLoadHelper.loadPicture(ivCover, String.format("%s.%s", comics.image.path,
                    comics.image.extension))
        } else {
            ImageLoadHelper.loadPictureByDrawable(ivCover, R.drawable.image_error_marvel_logo)
        }
    }

    override fun hideProgress() {
        progressBar.visibility = View.GONE
    }

    override fun showProgress() {
        progressBar.visibility = View.VISIBLE
    }

    private fun initViews() {
        findViews()
        supportActionBar(toolbar)
        setBackArrow(toolbar)
        collapsingToolbar.title = intent.getStringExtra(NAME_KEY)
    }

    private fun findViews() {
        collapsingToolbar = findViewById(R.id.ct_comics)
        toolbar = findViewById(R.id.tb_comics)
        ivCover = findViewById(R.id.iv_comics)
        tvDescription = findViewById(R.id.tv_description)
        tvPrice = findViewById(R.id.tv_price)
        tvPages = findViewById(R.id.tv_pages)
        progressBar = findViewById(R.id.progress_loader)
    }

    companion object {

        @JvmStatic
        fun start(activity: Activity, comics: Comics) {
            val intent = Intent(activity, ComicsActivity::class.java)
            intent.putExtra(NAME_KEY, comics.name)
            intent.putExtra(ID_KEY, comics.id)
            activity.startActivity(intent)
        }
    }
}
